### PR TITLE
Added endpoint caching API

### PR DESCRIPTION
**Pull request by joonasakoski**
- Added new IEndPointCache API
- DefaultEndpointCache uses internally IApiCache which uses IDistributionCache. IDistributionCache is implemented as DistricbutedMemoryCache

### CODE REVIEW

File: src/Weikio.ApiFramework.Abstractions/IApiCache.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        string GetOrCreateString(Endpoint endpoint, string key, Func<string> getString);
+        Task<string> GetOrCreateStringAsync(Endpoint endpoint, string key, Func<Task<string>> getString);
+        string GetString(Endpoint endpoint, string key);
+        void SetString(Endpoint endpoint, string key, string value);
+
+        byte[] GetOrCreateObject(Endpoint endpoint, string key, Func<byte[]> getObject);
+        Task<byte[]> GetOrCreateObjectAsync(Endpoint endpoint, string key, Func<Task<byte[]>> getObject);
+        byte[] GetObject(Endpoint endpoint, string key);
+        void SetObject(Endpoint endpoint, string key, byte[] value);
```
Comment (tsilvonen) [inline src/Weikio.ApiFramework.Abstractions/IApiCache.cs:18]: These methods could be named like this: `GetOrCreateData`, `GetData`, `SetData` etc.

File: src/Weikio.ApiFramework.Abstractions/IEndpointCache.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        string GetOrCreateString(string key, Func<string> getString);
+        Task<string> GetOrCreateStringAsync(string key, Func<Task<string>> getString);
+        string GetString(string key);
+        void SetString(string key, string value);
+
+        byte[] GetOrCreateObject(string key, Func<byte[]> getObject);
+        Task<byte[]> GetOrCreateObjectAsync(string key, Func<Task<byte[]>> getObject);
+        byte[] GetObject(string key);
+        void SetObject(string key, byte[] value);
```
Comment (tsilvonen) [inline src/Weikio.ApiFramework.Abstractions/IEndpointCache.cs:19]: These methods could be named like this: GetOrCreateData, GetData, SetData etc.

File: src/Weikio.ApiFramework.Core/Cache/ApiCacheOptions.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Weikio.ApiFramework.Abstractions;
+
+namespace Weikio.ApiFramework.Core.Cache
+{
+    public class ApiCacheOptions
+    {
+        public int ExpirationTimeInSeconds { get; set; } = 60;
```
Comment (tsilvonen) [inline src/Weikio.ApiFramework.Core/Cache/ApiCacheOptions.cs:10]: Rename this to `ExpirationTime` and change the type to `TimeSpan`.

File: src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                item = getString();
+                _distributedCache.SetString(cacheKey, item);
+            }
+            return item;
+        }
+
+        public async Task<string> GetOrCreateStringAsync(Endpoint endpoint, string key, Func<Task<string>> getString)
+        {
+            var cacheKey = _apiCacheOptions.GetKey(endpoint, _serviceProvider, key);
+            var item = _distributedCache.GetString(cacheKey.ToString());
```
Comment (tsilvonen) [inline src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs:41]: I think we should use `_distributedCache.GetStringAsync` here?

File: src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                item = getObject();
+                _distributedCache.Set(cacheKey, item);
+            }
+            return item;
+        }
+
+        public async Task<byte[]> GetOrCreateObjectAsync(Endpoint endpoint, string key, Func<Task<byte[]>> getObject)
+        {
+            var cacheKey = _apiCacheOptions.GetKey(endpoint, _serviceProvider, key);
+            var item = _distributedCache.Get(cacheKey.ToString());
```
Comment (tsilvonen) [inline src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs:65]: I think we should use `_distributedCache.GetAsync` here?

File: src/Weikio.ApiFramework.Core/Cache/DefaultEndpointCache.cs [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+        public DefaultEndpointCache(IApiCache apiCache, IHttpContextAccessor contextAccessor)
+        {
+            _apiCache = apiCache;
+            _contextAccessor = contextAccessor;
+        }
+
+        public string GetOrCreateString(string key, Func<string> getString)
+        {
+            var endpoint = _contextAccessor.HttpContext.GetEndpointMetadata();
```
Comment (tsilvonen) [inline src/Weikio.ApiFramework.Core/Cache/DefaultEndpointCache.cs:24]: Create a new private method `GetEndpoint` and throw `InvalidOperationException` if `_contextAccessor.HttpContext` is null. Use the new method in all other publi methods in this class.

File: tests/Apis/HelloWorld/HelloWorldCacheApi.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        private object GetTestObject(string name)
+        {
+            var testObject = new TestObject()
+            {
+                Name = name
+            };
+            return testObject;
+        }
```
Comment (tsilvonen) [inline tests/Apis/HelloWorld/HelloWorldCacheApi.cs:36]: This is not used anywhere. Also `TestObject` will be unused if this method is removed.

**Review by tsilvonen — CHANGES_REQUESTED** [review]

File: src/Weikio.ApiFramework.Core/Cache/ApiCacheOptions.cs (hunk 3 above)
Comment (mikoskinen) [reply src/Weikio.ApiFramework.Core/Cache/ApiCacheOptions.cs:10]: I think the idea here was that using seconds is easier than TimeSpans if we are configuring things through appsettings.json or some other json-file.

But yep, this should be changed. TimeSpan can be written as 00:00:60 in appsettings.

File: src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        public byte[] GetOrCreateObject(Endpoint endpoint, string key, Func<byte[]> getObject)
+        {
+            var cacheKey = _apiCacheOptions.GetKey(endpoint, _serviceProvider, key);
+            var item = _distributedCache.Get(cacheKey.ToString());
+            if (item == null)
+            {
+                item = getObject();
+                _distributedCache.Set(cacheKey, item);
+            }
+            return item;
```
Comment (mikoskinen) [inline src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs:59]: Should have an empty line before return (as defined in Editorconfig).

File: src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        public async Task<byte[]> GetOrCreateObjectAsync(Endpoint endpoint, string key, Func<Task<byte[]>> getObject)
+        {
+            var cacheKey = _apiCacheOptions.GetKey(endpoint, _serviceProvider, key);
+            var item = _distributedCache.Get(cacheKey.ToString());
+            if (item == null)
+            {
+                item = await getObject();
+                _distributedCache.Set(cacheKey, item);
+            }
+            return item;
```
Comment (mikoskinen) [inline src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs:71]: Should have an empty line before return (as defined in Editorconfig).

File: src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+        {
+            var cacheKey = _apiCacheOptions.GetKey(endpoint, _serviceProvider, key);
+            var options = GetEntryOptions();
+            _distributedCache.SetString(cacheKey, value, options);
+        }
+
+        public byte[] GetObject(Endpoint endpoint, string key)
+        {
+            var cacheKey = _apiCacheOptions.GetKey(endpoint, _serviceProvider, key);
+            return _distributedCache.Get(cacheKey.ToString());
```
Comment (mikoskinen) [inline src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs:90]: Should have an empty line before return (as defined in Editorconfig).

File: src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+                item = await getObject();
+                _distributedCache.Set(cacheKey, item);
+            }
+            return item;
+        }
+
+        public string GetString(Endpoint endpoint, string key)
+        {
+            var cacheKey = _apiCacheOptions.GetKey(endpoint, _serviceProvider, key);
+            return _distributedCache.GetString(cacheKey.ToString());
```
Comment (mikoskinen) [inline src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs:77]: Should have an empty line before return (as defined in Editorconfig).

File: src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        private DistributedCacheEntryOptions GetEntryOptions()
+        {
+            var options = new DistributedCacheEntryOptions();
+            if (_apiCacheOptions.ExpirationTimeInSeconds > 0)
+            {
+                options.AbsoluteExpirationRelativeToNow = TimeSpan.FromSeconds(_apiCacheOptions.ExpirationTimeInSeconds);
+            }
+            return options;
```
Comment (mikoskinen) [inline src/Weikio.ApiFramework.Core/Cache/DefaultApiCache.cs:107]: Should have an empty line before return (as defined in Editorconfig).

File: src/Weikio.ApiFramework.Core/Extensions/ServiceCollectionExtensions.cs [hunk 13]
```diff
@@ -186,6 +187,10 @@ public static class ServiceCollectionExtensions
                 mvcOptions.Filters.Add(filter);
             });
 
+            services.AddDistributedMemoryCache();
```
Comment (mikoskinen) [inline src/Weikio.ApiFramework.Core/Extensions/ServiceCollectionExtensions.cs:190]: The DistributedMemoryCache should be added only if it isn't already configured in the app. Make sure (by checking the source code of AddDistributedMemoryCache) that this is skipped if configuration has already happened before this line.

File: tests/Apis/HelloWorld/HelloWorld.csproj [hunk 14]
```diff
@@ -8,4 +8,8 @@
     <FrameworkReference Include="Microsoft.AspNetCore.App" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\src\Weikio.ApiFramework.Abstractions\Weikio.ApiFramework.Abstractions.csproj" />
```
Comment (mikoskinen) [inline tests/Apis/HelloWorld/HelloWorld.csproj:12]: Should reference SDK, not the Abstractions.

**Review by mikoskinen — COMMENTED** [review]
Make sure that your IDE is using the editorconfig.

File: src/Weikio.ApiFramework.Core/Cache/ApiCacheOptions.cs (hunk 3 above)
Comment (tsilvonen) [reply src/Weikio.ApiFramework.Core/Cache/ApiCacheOptions.cs]: Done.

File: tests/Apis/HelloWorld/HelloWorldCacheApi.cs (hunk 7 above)
Comment (tsilvonen) [reply tests/Apis/HelloWorld/HelloWorldCacheApi.cs]: Removed.

File: src/Weikio.ApiFramework.Core/Cache/DefaultEndpointCache.cs (hunk 6 above)
Comment (tsilvonen) [reply src/Weikio.ApiFramework.Core/Cache/DefaultEndpointCache.cs]: Done.

File: src/Weikio.ApiFramework.Core/Extensions/ServiceCollectionExtensions.cs (hunk 13 above)
Comment (tsilvonen) [reply src/Weikio.ApiFramework.Core/Extensions/ServiceCollectionExtensions.cs]: `AddDistributedMemoryCache()` uses `TryAdd` internally but it's not clear to the reader without checking the source code. 

Made this more explicit by wrapping the statement in if-clause which checks the existence of service type `IDistributedCache`:

```
if (!services.Any(s => s.ServiceType == typeof(IDistributedCache)))
{ 
    services.AddDistributedMemoryCache();
}
```

File: tests/Apis/HelloWorld/HelloWorld.csproj (hunk 14 above)
Comment (tsilvonen) [reply tests/Apis/HelloWorld/HelloWorld.csproj]: Done.

File: src/Weikio.ApiFramework.Abstractions/IApiCache.cs (hunk 1 above)
Comment (tsilvonen) [reply src/Weikio.ApiFramework.Abstractions/IApiCache.cs]: Now `IApiCache` defines only the bare minimum of operations. Basically only `GetData`, `GetDataAsync`, `SetData` and `SetDataAsync` are defined. 

Other operations (`GetOrCreate`, `GetString`, etc.) are defined in `IApiCacheExtensions`.

File: src/Weikio.ApiFramework.Abstractions/IEndpointCache.cs (hunk 2 above)
Comment (tsilvonen) [reply src/Weikio.ApiFramework.Abstractions/IEndpointCache.cs]: Now `IEndpointCache` defines only the bare minimum of operations. Basically only `GetData`, `GetDataAsync`, `SetData` and `SetDataAsync` are defined. 

Other operations (`GetOrCreate`, `GetString`, etc.) are defined in `IEndpointCacheExtensions`.

**Review by tsilvonen — APPROVED** [review]

**Review by mikoskinen — APPROVED** [review]

**Pull request merged**